pkg/router/kubernetes_default.go

```diff
@@ -112,15 +112,34 @@ func (c *KubernetesDefaultRouter) reconcileService(canary *flaggerv1.Canary, nam
 		svcSpec.Ports = append(svcSpec.Ports, cp)
 	}
 
+	if metadata == nil {
+		metadata = &flaggerv1.CustomMetadata{}
+	}
+
+	if metadata.Labels == nil {
+		metadata.Labels = make(map[string]string)
+	}
+	if metadata.Annotations == nil {
+		metadata.Annotations = make(map[string]string)
+	}
+
+	metadata.Labels[c.labelSelector] = name
+	for k, v := range c.annotations {
+		metadata.Annotations[k] = v
+	}
+
+	c.logger.With("canary", fmt.Sprintf("%s.%s", canary.Name, canary.Namespace)).
+		Debugw(fmt.Sprintf("Creating Service %s", name), "metadata", metadata, "service_configuration", canary.Spec.Service)
+
 	// create service if it doesn't exists
 	svc, err := c.kubeClient.CoreV1().Services(canary.Namespace).Get(name, metav1.GetOptions{})
 	if errors.IsNotFound(err) {
 		svc = &corev1.Service{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:        name,
 				Namespace:   canary.Namespace,
-				Labels:      map[string]string{c.labelSelector: name},
-				Annotations: c.annotations,
+				Labels:      metadata.Labels,
+				Annotations: metadata.Annotations,
```

Comment:
This only applies to new services but we should also cover updates as well. One can change the metadata in the canary spec or path the service in-cluster, on both cases we should reconcile with the canary spec.